Exclude bounded key selectors from result key selector after bound

denbezrukov opened the pull request:
We can use only objects for `createBoundSelector` otherwise it's meaningless and we have to think about this case. If binding is object we can extract all possible paths which we want to bound. These paths help us to remove bounded key selectors from result key selector. 

**NOTE 1**: If `baseKeySelector` has dependencies which aren't `createPropSelector` we can't exclude them although some of they can be bounded.

**NOTE 2**: This optimization works with conjunction `composingKeySelectorCreator` because all dependencies are flat and uniq.

Commit:
optimization:  exclude bounded key selectors from result key selector after bound

Commit:
make binding for every exotic prop selector

sgrishchenko commented on packages/reselect-utils/src/helpers.ts:
```diff
@@ -39,3 +39,62 @@ export const isDebugMode = () => debugMode;
+      }
+    }
+  }
+  return paths;
```
I propose to implement `getObjectPaths` this way:
```js
getObjectPaths = (value, parentPath = []) => {
  if (!isObject(value)) return [parentPath];
 
  return Object.keys(value).reduce((result, key) => {
    const paths = getObjectPaths(
      value[key], 
      parentPath.concat([key])
    );
    return result.concat(paths)
  }, []);
}
```

sgrishchenko commented on packages/reselect-utils/src/createPropSelector.ts:
```diff
@@ -6,7 +6,9 @@ import {
-export const isPropSelector = (selector: unknown) =>
+export const isPropSelector = (
+  selector: unknown,
+): selector is { path: (string | number)[] } =>
```
I propose to create `type Path = string[]` because `Object.keys` always returns strings.

sgrishchenko commented on packages/reselect-utils/src/__tests__/createBoundSelector.test.ts:
```diff
@@ -185,5 +190,195 @@ describe('createAdaptedSelector', () => {
+        value2: 2,
+        value3: 5,
+        value4: 6,
+      }) as unknown;
```
I think we can simplify this test:
```js
const key: unknown = keySelector(commonState, {
        value2: 2,
        value3: 3,
      });
```

sgrishchenko commented on packages/reselect-utils/src/__tests__/createBoundSelector.test.ts:
```diff
@@ -185,5 +190,195 @@ describe('createAdaptedSelector', () => {
+    expect(isComposedKeySelector(keySelector)).toBeFalsy();
+  });
+
+  test('should return composed key selector if result key sector has more than one dependency', () => {
```
I think you meant key se**le**ctor

sgrishchenko commented on packages/reselect-utils/src/__tests__/createBoundSelector.test.ts:
```diff
@@ -185,5 +190,195 @@ describe('createAdaptedSelector', () => {
+    expect(isComposedKeySelector(keySelector)).toBeTruthy();
+  });
+
+  test('should bound not props selector without exclude', () => {
```
it wold be better to say `non prop selector`

sgrishchenko commented on packages/reselect-utils/src/__tests__/helpers.test.ts:
```diff
@@ -0,0 +1,69 @@
+      ['key1', 'key6', 'key8'],
+      ['key1', 'key4', 'key5'],
+      ['key1', 'key2', 'key3'],
+    ]);
```
May be we may add some path with different length to show that paths lengths are not fixed

Commit:
Fixes after review

- Simplify createBoundSelector tests.
- Extract path type
- Reimplement getObjectPaths
